org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MavenProjectFacade.java

```diff
@@ -532,7 +535,10 @@ public List<MojoExecution> getExecutionPlan(String lifecycle, IProgressMonitor m
 
   @Override
   public IMaven getMaven() {
-    // TODO: if project has Maven customization (such as a .mvn, MAVEN_OPTS...)
-    return MavenPlugin.getMaven();
+    if(maven == null) {
+      File basedir = pomFile != null ? (pomFile.isDirectory() ? pomFile : pomFile.getParentFile()) : null;
+      maven = ((MavenImpl) MavenPlugin.getMaven()).cloneForBasedir(basedir);
```

Comment:
Actually the search for the `.mvn` folder should happen here and we should then get a MavenImpl per `.mvn` folder (that could be shared by multiple projects).